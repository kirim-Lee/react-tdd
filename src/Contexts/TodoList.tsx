import { createContext, useEffect, useState } from 'react';

interface IContext {
  readonly todoList: string[];
  readonly addTodo: (toDo: string) => void;
  readonly deleteTodo: (index: number) => void;
}

const TodoListContext = createContext<IContext>({
  todoList: [],
  addTodo: () => {},
  deleteTodo: () => {},
});

interface IProps {
  children: JSX.Element | JSX.Element[];
}

const TODO_LIST = 'TodoList';

const TodoListProvider = ({ children }: IProps): JSX.Element => {
  const [todoList, setTodoList] = useState<string[]>([]);

  const addTodo = (toDo: string) => {
    if (!toDo) return;

    const newList = [...todoList, toDo];
    localStorage.setItem(TODO_LIST, JSON.stringify(newList));
    setTodoList(newList);
  };

  const deleteTodo = (index: number) => {
    let list = [...todoList];
    list.splice(index, 1);
    localStorage.setItem(TODO_LIST, JSON.stringify(list));
    setTodoList(list);
  };

  useEffect(() => {
    const list = localStorage.getItem(TODO_LIST);
    if (list) {
      setTodoList(JSON.parse(list));
    }
  }, []);

  return (
    <TodoListContext.Provider value={{ todoList, addTodo, deleteTodo }}>
      {children}
    </TodoListContext.Provider>
  );
};

export { TodoListContext, TodoListProvider };
