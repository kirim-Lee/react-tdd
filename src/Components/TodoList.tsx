import React, { useContext, useState } from 'react';
import styled from 'styled-components';
import { TodoItem } from './TodoItem';
import { TodoListContext } from '../Contexts/TodoList';

const Container = styled.div`
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #bdbdbd;
`;

export const TodoList = () => {
  const { todoList, deleteTodo } = useContext(TodoListContext);

  return (
    <Container data-testid="toDoList">
      {todoList.map((item, index) => (
        <TodoItem
          key={item}
          id={index}
          label={item}
          onDelete={() => deleteTodo(index)}
        />
      ))}
    </Container>
  );
};
