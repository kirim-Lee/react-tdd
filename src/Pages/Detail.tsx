import { useNavigate, useParams } from 'react-router-dom';
import { Button, Container } from '../Components';
import { useContext, useMemo } from 'react';
import { TodoListContext } from '../Contexts/TodoList';
import styled from 'styled-components';

const ToDo = styled.div`
  min-width: 350px;
  height: 350px;
  overflow-y: auto;
  border: 1px solid #bdbdbd;
  margin-bottom: 20px;
  padding: 10px;
`;

const Detail = () => {
  const navigate = useNavigate();
  const params = useParams<{ id: string }>();
  const { todoList, deleteTodo } = useContext(TodoListContext);

  const id = useMemo(() => {
    return Number(params.id);
  }, [params?.id]);

  const handleDelete = () => {
    deleteTodo(id);
    navigate(-1);
  };

  return (
    <Container>
      <ToDo>{todoList[id]}</ToDo>
      <Button
        label="삭제"
        backgroundColor="#ff1744"
        hoverColor="#f01440"
        onClick={handleDelete}
      />
    </Container>
  );
};

export default Detail;
