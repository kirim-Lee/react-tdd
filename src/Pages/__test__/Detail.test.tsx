import {
  MemoryRouter,
  Route,
  RouterProvider,
  Routes,
  createMemoryRouter,
  useLocation,
} from 'react-router-dom';

import { fireEvent, render, screen } from '@testing-library/react';
import { TodoListProvider } from '../../Contexts/TodoList';
import Detail from '../Detail';

const TODO_LIST = 'TodoList';

beforeEach(() => {
  localStorage.clear();
});

describe('Detail page', () => {
  it('렌더링 테스트', () => {
    const router = createMemoryRouter(
      [{ path: '/detail/:id', element: <Detail /> }],
      { initialEntries: ['/detail/1'] }
    );

    localStorage.setItem(TODO_LIST, JSON.stringify(['Todo 1', 'Todo 2']));

    const { container } = render(
      <TodoListProvider>
        <RouterProvider router={router} />
      </TodoListProvider>
    );

    const todoItem = screen.getByText('Todo 2');
    expect(todoItem).toBeInTheDocument();

    const button = screen.getByText('삭제');
    expect(button).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('삭제 버튼 클릭시 삭제 잘되는지 확인, 홈으로 이동되는지 확인', () => {
    const TestComponent = () => {
      const location = useLocation();
      return <div>{location.pathname}</div>;
    };

    localStorage.setItem(TODO_LIST, JSON.stringify(['Todo1', 'Todo2']));

    render(
      <TodoListProvider>
        <MemoryRouter initialEntries={['/', '/detail/0']} initialIndex={1}>
          <TestComponent />
          <Routes>
            <Route path="/detail/:id" element={<Detail />} />
          </Routes>
        </MemoryRouter>
      </TodoListProvider>
    );

    const url = screen.getByText('/detail/0');
    expect(url).toBeInTheDocument();

    const todoItem = screen.getByText('Todo1');
    expect(todoItem).toBeInTheDocument();

    const button = screen.getByText('삭제');
    expect(button).toBeInTheDocument();

    fireEvent.click(button);

    expect(url.textContent).toBe('/');
    expect(todoItem).not.toBeInTheDocument();
    expect(button).not.toBeInTheDocument();

    expect(localStorage.getItem(TODO_LIST)).toBe(JSON.stringify(['Todo2']));
  });
});
