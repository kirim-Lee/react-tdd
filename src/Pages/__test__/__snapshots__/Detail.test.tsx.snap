// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detail page 렌더링 테스트 1`] = `
<div>
  <div
    class="sc-bcXHqe jnXeEi"
  >
    <div
      class="sc-kDvujY kvPBTJ"
    >
      Todo 2
    </div>
    <button
      class="sc-eDvSVe kbfjFZ"
    >
      <div
        class="sc-jSUZER jTTUoN"
      >
        삭제
      </div>
    </button>
  </div>
</div>
`;
